Topic step15

CraigCCC opened the pull request:
- Add task status enum (pending, processing, done)
- Add search feature (can type title or select status)  in index page
- Add search index (title)
- Add search feature (model spec & feature spec)
- Deploying this branch (Topic_step15) to Heroku

CraigCCC committed:
Add search feature
- Add search logic to products#index and edit after search can sort
- Add search scope
- Add search field view
- Add search link I18n

CraigCCC committed:
Add index to task table name column

CraigCCC committed:
Add search feature(status)
- Add search_status logic to products#index and edit after search can sort
- Add search_status scope
- Add search_status option_select view
- Enabled Rubocop Metrics/AbcSize

CraigCCC committed:
Add model test(search)
- Edit tasks_controller, add sort_by_created_at to index search result
- Modify FactoryBot, use trait
- Add 'Search Task' model test

CraigCCC committed:
Add feature test(search and status)
- Add 'Search feature' by title, status, title and status

chaowudev commented on app/models/task.rb:
```diff
@@ -10,6 +10,11 @@ class Task < ApplicationRecord
   # scope
   scope :sort_by_created_at, -> { order('created_at DESC') }
   scope :sort_by_column, ->(order_by, direction) { order("#{order_by} #{direction}") }
+  scope :search_like, ->(search) { search.present? ? where('title ILIKE ?', "%#{search}%") : all }
```
`scope :search_title, -> (title)`

chaowudev commented on app/models/task.rb:
```diff
@@ -10,6 +10,11 @@ class Task < ApplicationRecord
   scope :sort_by_created_at, -> { order('created_at DESC') }
   scope :sort_by_column, ->(order_by, direction) { order("#{order_by} #{direction}") }
+  scope :search_like, ->(search) { search.present? ? where('title ILIKE ?', "%#{search}%") : all }
+  scope :search_status, ->(search_status) { search_status.present? ? where('status = ?', search_status) : all }
```
`scope :search_stauts, -> (status)`

chaowudev commented on app/controllers/tasks_controller.rb:
```diff
@@ -2,9 +2,16 @@ class TasksController < ApplicationController
-    return @tasks = Task.sort_by_column(params[:order_by], params[:direction]) if params[:order_by]
-
-    @tasks = Task.sort_by_created_at
+    if params[:search] || params[:search_status]
```
把這邊的判斷拉出去做一個 search method

CraigCCC replied:
已經更改成一個 private method

chaowudev commented on app/models/task.rb:
```diff
@@ -10,6 +10,11 @@ class Task < ApplicationRecord
+  scope :search_status, ->(search_status) { search_status.present? ? where('status = ?', search_status) : all }
+
+  # Constant
+  SEARCH_STATUS_CLASS = [[I18n.t('tasks.status.pending'), 0], [I18n.t('tasks.status.processing'), 1], [I18n.t('tasks.status.done'), 2]].freeze
```
通常會把常數放在這個檔案的最上方，原因在於易於閱讀，如果在一個檔案中看到常數也能直覺性的拉到最上方去查看。
```
class Task < ApplicationRecord
  # constant
  SEARCH_STATUS_CLASS = [[I18n.t('tasks.status.pending'), 0], [I18n.t('tasks.status.processing'), 1], [I18n.t('tasks.status.done'), 2]].freeze

  # validation
  validates :title, :status, :priority, :start_at, :end_at, presence: true
  validates :title, length: { maximum: 50 }
  # ...
```

CraigCCC replied:
這裡的 constant 我將它刪除了，改用 Helper 一個 method 去做 status 的 select option
因為如果按照 rubocop 設定要將 constant freeze ，再切換 I18n 時，選項並不會更改。

chaowudev commented on spec/features/task_spec.rb:
```diff
@@ -93,4 +120,18 @@ def expect_task_info_equal(task)
+    end
+    click_button '搜尋'
+    page.all('.task_title').each do |target|
+      expect(target).to have_content(title)
```
params 命名
`expect(task).to have_content(title)`
可改可不改

chaowudev commented on spec/features/task_spec.rb:
```diff
@@ -93,4 +120,18 @@ def expect_task_info_equal(task)
+      expect(target).to have_content(title)
+    end
+    page.all('.task_status').each do |target|
+      expect(target).to have_content(status)
```
同上

chaowudev commented on spec/features/task_spec.rb:
```diff
@@ -85,6 +85,33 @@
 
+  describe 'Search feature', js: true do
+    scenario 'when search by title' do
+      FactoryBot.create(:task)
```
可以看一下 `before` 和 `let` 的用法

CraigCCC replied:
已更改

chaowudev commented on spec/models/task_spec.rb:
```diff
@@ -33,4 +34,39 @@
+
+  private
+
+  def search_by_title_and_status(target_task, title, status, enum)
```
可以給預設值 nil
`def search_by_title_and_status(task, title = nil, status = nil, enum = nil)`
這樣在上面使用方法時，最後兩個就不用特別給參數 `search_by_title_and_status(target_task, 'Hello World', '', '',)`

CraigCCC replied:
如果有四個引數要填，中間兩個也可以特別不給參數嗎？
還是說這種情況就是要給 nil 了

CraigCCC committed:
Fix code
- Refactoring tasks#index by use private method index_sort_or_search
- Setup TasksHelper method status_options to solve when using I18n can't transform options
 - remove constant SEARCH_STATUS_CLASS
- Spec/
  - Use let to replace FactoryBot.create
  - Setup default argument nil to replace empty string

chaowudev commented on spec/features/task_spec.rb:
```diff
@@ -93,4 +125,18 @@ def expect_task_info_equal(task)
+      fill_in 'search', with: title
+      select status, from: 'search_status' if status.present?
+    end
+    click_button '搜尋'
```
把填資料的行為跟期望行為拆成兩個方法，這樣在閱讀上比較不用上下對照就知道這段測試在幹嘛
```ruby
scenario 'when search by title' do
    task
    task_title

    visit root_path
    search_by_title_and_status('Hello World')
    click_button '搜尋'

    expect_search_result_with('Hello World')
end

def search_by_title_and_status(title = nil, status = nil)
    within('form') do
      fill_in 'search', with: title
      select status, from: 'search_status' if status.present?
    end
end

def expect_search_result_with(title = nil, status = nil)
    page.all('.task_title').each do |task|
      expect(task).to have_content(title)
    end
    page.all('.task_status').each do |task|
      expect(task).to have_content(status)
    end
end
```

chaowudev commented on spec/models/task_spec.rb:
```diff
@@ -33,4 +34,39 @@
+      processing_task = FactoryBot.create(:task, :processing)
+      done_task = FactoryBot.create(:task, :done)
+
+      search_by_title_and_status(pending_task, nil, 'pending', '0')
```
這邊不會需要測到 title 的話，你可以調一下 method 參數的順序
`def search_by_title_and_status(task, status = nil, title = nil, enum = nil)`
這樣連 nil 都不用寫了，閱讀上也不用想那個 nil 是哪個欄位的值是空的
但比較建議是用 hash 來寫方法的參數，這樣就不會有順序的困擾了

CraigCCC committed:
Fix code
- Rspec/features/task_spec.rb
  - Modify describe

chaowudev reviewed (approved):
LGTM
